diff --git a/include/groundhog/traj_block_format.h b/include/groundhog/traj_block_format.h
index dc2aee2..35567e4 100644
--- a/include/groundhog/traj_block_format.h
+++ b/include/groundhog/traj_block_format.h
@@ -21,6 +21,8 @@ struct seg_meta {
     int time_max;
     int seg_offset;
     int seg_size;
+    void* oid_filter;
+    int oid_filter_size;    // the unit is byte (8 bits)
 };
 
 struct traj_block_header {
